### PR TITLE
allowUsingHttp2

### PR DESCRIPTION
**Pull request by xinlian12**
* Added support to enable http2 with system property `COSMOS.HTTP2_ENABLED` and system variable `COSMOS_HTTP2_ENABLED`
* Added support to allow changing http2 max connection pool size with system property `COSMOS.HTTP2_MAX_CONNECTION_POOL_SIZE` and system variable `COSMOS_HTTP2_MAX_CONNECTION_POOL_SIZE`
* Added support to allow changing http2 max connection pool size with system property `COSMOS.HTTP2_MIN_CONNECTION_POOL_SIZE` and system variable `COSMOS_HTTP2_MIN_CONNECTION_POOL_SIZE`
* Added support to allow changing http2 maximum number of the concurrent streams that can be opened to the remote peer with system property `COSMOS.HTTP2_MAX_CONCURRENT_STREAMS` and system variable `COSMOS_HTTP2_MAX_CONCURRENT_STREAMS`

### CODE REVIEW

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/http/Http2ResponseHeaderCleanerHandler.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+            Http2Headers headers = headersFrame.headers();
+
+            headers.forEach(entry -> {
+                CharSequence key = entry.getKey();
+                CharSequence value = entry.getValue();
+
+                // Check for leading whitespace or other prohibited characters
+                if (StringUtils.isNotEmpty(value) && (value.charAt(0) == ' ' || value.charAt(value.length() - 1) == ' ')) {
+                    // Clean up the header value by trimming or handling as needed
+                    logger.warn("There are extra white space for key {} with value {}", key, value);
```
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/http/Http2ResponseHeaderCleanerHandler.java:31]: Given taht this happens for each request now - maybe make this DEBUG level tracing for now?

Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/http/Http2ResponseHeaderCleanerHandler.java:31]: make sense - will change in next iteration

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/http/HttpClient.java [hunk 2]
```diff
@@ -59,6 +59,8 @@ static HttpClient createFixed(HttpClientConfig httpClientConfig) {
         fixedConnectionProviderBuilder.pendingAcquireTimeout(connectionAcquireTimeout);
         fixedConnectionProviderBuilder.maxIdleTime(maxIdleConnectionTimeoutInMillis);
 
+        // TODO[Http2]: config Http2AllocationStrategy (maxConnections & maxConcurrentStreams)
```
[the remark this answers is not in the document]
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/http/HttpClient.java:62]: yea want to wire it up when adding the public API change

**Review by FabianMeiswinkel — APPROVED** [review]
LGTM - Thanks!

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/http/Http2ResponseHeaderCleanerHandler.java (hunk 1 above)
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/http/Http2ResponseHeaderCleanerHandler.java:31]: changed to trace level

**Review by FabianMeiswinkel — APPROVED** [review]
LGTM - Thanks!

**Review by kushagraThapar — APPROVED** [review]
Looks good overall, thanks @xinlian12 
I added few suggestions!

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/http/HttpClient.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/http/HttpClient.java]: wired up with Configs

File: sdk/cosmos/azure-cosmos/CHANGELOG.md [hunk 3]
```diff
@@ -9,6 +9,10 @@
 #### Bugs Fixed
 
 #### Other Changes
+* Added support to enable http2 for gateway mode with system property `COSMOS.HTTP2_ENABLED` and system variable `COSMOS_HTTP2_ENABLED`. - [PR 42947](https://github.com/Azure/azure-sdk-for-java/pull/42947)
```
[the remark this answers is not in the document]
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/CHANGELOG.md:12]: plan to leave them here as in the next PR, these configs will become public API, but let me know your thoughts

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/http/Http2ResponseHeaderCleanerHandler.java [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+        if (msg instanceof Http2HeadersFrame) {
+            Http2HeadersFrame headersFrame = (Http2HeadersFrame) msg;
+            Http2Headers headers = headersFrame.headers();
+
+            headers.forEach(entry -> {
+                CharSequence key = entry.getKey();
+                CharSequence value = entry.getValue();
+
+                // Check for leading whitespace or other prohibited characters
+                if (StringUtils.isNotEmpty(value) && (value.charAt(0) == ' ' || value.charAt(value.length() - 1) == ' ')) {
```
[the remark this answers is not in the document]
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/http/Http2ResponseHeaderCleanerHandler.java]: changed to only check `x-ms-serviceversion` header

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/http/HttpClient.java [hunk 5]
```diff
@@ -45,6 +47,16 @@ static HttpClient createFixed(HttpClientConfig httpClientConfig) {
         fixedConnectionProviderBuilder.pendingAcquireTimeout(httpClientConfig.getConnectionAcquireTimeout());
         fixedConnectionProviderBuilder.maxIdleTime(httpClientConfig.getMaxIdleConnectionTimeout());
 
+        if (httpClientConfig.isHttp2Enabled()) {
+            fixedConnectionProviderBuilder.allocationStrategy(
+                Http2AllocationStrategy.builder()
+                    .maxConnections(Configs.getHttp2MaxConnectionPoolSize())
```
[the remark this answers is not in the document]
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/http/HttpClient.java]: Changed to use Http2ConnectionConfig internally

**Comment by xinlian12** [comment]
Flaky tests: readFeedDocumentsAfterSplit

**Pull request merged**